src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -0,0 +1,515 @@
+        {
+          if (HappyTreeFriends::isLeaf(_bvh, left_child))
+          {
+            float const candidate_dist = _metric(
+                leaf_permutation_i,
+                HappyTreeFriends::getLeafPermutationIndex(_bvh, left_child),
+                distance_left);
+            WeightedEdge const candidate_edge{i, left_child, candidate_dist};
+            if (candidate_edge < current_best)
+            {
+              current_best = candidate_edge;
+              Kokkos::atomic_min(&radius, candidate_dist);
+            }
+          }
+          else
+          {
+            traverse_left = true;
+          }
+        }
+
+        // Note: radius may have been already updated here from the left child
+        if (predicate(right_child) &&
+            (distance_right = distance(right_child)) <= radius)
+        {
+          if (HappyTreeFriends::isLeaf(_bvh, right_child))
+          {
+            float const candidate_dist = _metric(
+                leaf_permutation_i,
+                HappyTreeFriends::getLeafPermutationIndex(_bvh, right_child),
+                distance_right);
+            WeightedEdge const candidate_edge{i, right_child, candidate_dist};
+            if (candidate_edge < current_best)
+            {
+              current_best = candidate_edge;
+              Kokkos::atomic_min(&radius, candidate_dist);
+            }
+          }
+          else
+          {
+            traverse_right = true;
+          }
+        }
+      }
+
+      if (!traverse_left && !traverse_right)
+      {
+        node = *--stack_ptr;
+#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+        if (node != SENTINEL)
+        {
+          // This is a theoretically unnecessary duplication of distance
+          // calculation for stack nodes. However, for Cuda it's better than
+          // putting the distances in stack.
+          distance_node = distance(node);
+        }
+#else
+        distance_node = *--stack_distance_ptr;
+#endif
+      }
+      else
+      {
+        node = (traverse_left &&
+                (distance_left <= distance_right || !traverse_right))
+                   ? left_child
+                   : right_child;
+        distance_node = (node == left_child ? distance_left : distance_right);
+        if (traverse_left && traverse_right)
+        {
+          *stack_ptr++ = (node == left_child ? right_child : left_child);
+#if !defined(__CUDA_ARCH__) && !defined(__HIP_DEVICE_COMPILE__)
+          *stack_distance_ptr++ =
+              (node == left_child ? distance_right : distance_left);
+#endif
+        }
+      }
+    } while (node != SENTINEL);
+
+    // This check is only here to reduce hammering the atomics for large
+    // components. Otherwise, for a large number of points and a small number of
+    // components it becomes extremely expensive.
+    auto &component_edge = _edges(component - n + 1);
+    if (current_best < component_edge)
```

Comment:
Note: this check still allows occasional unset edges in (i.e., `current_best = (i, -1, inf)`). Printing stuff out shows `component_edge` being `(i, j, inf)` with `i` and `j` being positive. I think this is because we don't read it atomically, so another thread may modify the `component_edge` weight after this one read indices but before this thread read in weight.

The overall result is still correct, though.

Reply:
```suggestion
    if (current_best.weight < inf)
```
or 
```suggestion
    if (current_best.weight < inf && current_best.weight <= component_edge.weight)
```